### PR TITLE
Feature/csi 3075 deduplication gap

### PR DESCRIPTION
**Pull request by ArbelNathan**

### CODE REVIEW

File: controller/array_action/config.py [hunk 1]
```diff
@@ -1,10 +1,12 @@
...
-SPACE_EFFICIENCY_THIN = 'thin'
-SPACE_EFFICIENCY_COMPRESSED = 'compressed'
-SPACE_EFFICIENCY_DEDUPLICATED = 'deduplicated'
-SPACE_EFFICIENCY_THICK = 'thick'
-SPACE_EFFICIENCY_NONE = 'none'
+SPACE_EFFICIENCY_THIN = "thin"
+SPACE_EFFICIENCY_COMPRESSED = "compressed"
+SPACE_EFFICIENCY_DEDUPLICATED = "deduplicated"
+SPACE_EFFICIENCY_DEDUPLICATED_COMPRESSED = "dedup_compressed"
+SPACE_EFFICIENCY_DEDUPLICATED_THIN = "dedup_thin"
```
Comment (oriyarde) [inline controller/array_action/config.py:7]:
```suggestion
SPACE_EFFICIENCY_THIN = "thin"
SPACE_EFFICIENCY_COMPRESSED = "compressed"
SPACE_EFFICIENCY_DEDUPLICATED = "deduplicated"
SPACE_EFFICIENCY_DEDUPLICATED_THIN = "dedup_thin"
SPACE_EFFICIENCY_DEDUPLICATED_COMPRESSED = "dedup_compressed"
```
to keep the thin-compressed order consistent

File: controller/array_action/array_mediator_svc.py [hunk 2]
```diff
@@ -87,9 +87,11 @@ def build_kwargs_from_parameters(space_efficiency, pool_name, volume_name,
             cli_kwargs.update({'thin': True})
         elif space_efficiency == config.SPACE_EFFICIENCY_COMPRESSED:
             cli_kwargs.update({'compressed': True})
-        elif space_efficiency == config.SPACE_EFFICIENCY_DEDUPLICATED:
-            cli_kwargs.update({'compressed': True, 'deduplicated': True})
-
+        elif space_efficiency == config.SPACE_EFFICIENCY_DEDUPLICATED_THIN:
+            cli_kwargs.update({'deduplicated': True, 'thin': True})
+        elif space_efficiency in [config.SPACE_EFFICIENCY_DEDUPLICATED,
+                                  config.SPACE_EFFICIENCY_DEDUPLICATED_COMPRESSED]:
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:93]:
```suggestion
        elif space_efficiency in (config.SPACE_EFFICIENCY_DEDUPLICATED,
                                  config.SPACE_EFFICIENCY_DEDUPLICATED_COMPRESSED):
```

Comment (ArbelNathan) [reply controller/array_action/array_mediator_svc.py]: done

File: controller/array_action/array_mediator_svc.py [hunk 3]
```diff
@@ -71,6 +71,21 @@ def is_warning_message(ex):
     return False
 
 
+def _get_array_space_efficiency(space_efficiency):
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:74]:
```suggestion
def _get_space_efficiency_kwargs(space_efficiency):
```

File: controller/array_action/array_mediator_svc.py [hunk 4]
```diff
@@ -80,16 +95,8 @@ def build_kwargs_from_parameters(space_efficiency, pool_name, volume_name,
...
-    if space_efficiency:
-        space_efficiency = space_efficiency.lower()
-        if space_efficiency == config.SPACE_EFFICIENCY_THIN:
-            cli_kwargs.update({'thin': True})
-        elif space_efficiency == config.SPACE_EFFICIENCY_COMPRESSED:
-            cli_kwargs.update({'compressed': True})
-        elif space_efficiency == config.SPACE_EFFICIENCY_DEDUPLICATED:
-            cli_kwargs.update({'compressed': True, 'deduplicated': True})
-
+    array_space_efficiency = _get_array_space_efficiency(space_efficiency)
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:98]:
```suggestion
    space_efficiency_kwargs = _get_space_efficiency_kwargs(space_efficiency)
```

File: controller/array_action/config.py (hunk 1 above)
Comment (oriyarde) [reply controller/array_action/config.py]: same [here](https://github.com/IBM/ibm-block-csi-driver/blob/d3de8b692bdca61e8f014aac186ba47bdd29c2fb/controller/array_action/array_mediator_svc.py#L361-L362)

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 5]
```diff
@@ -587,22 +597,34 @@ def test_validate_supported_space_efficiency_success(self):
         self.svc.validate_supported_space_efficiency(compressed_space_efficiency)
         deduplicated_space_efficiency = config.SPACE_EFFICIENCY_DEDUPLICATED
         self.svc.validate_supported_space_efficiency(deduplicated_space_efficiency)
+        deduplicated_thin_space_efficiency = config.SPACE_EFFICIENCY_DEDUPLICATED_THIN
+        self.svc.validate_supported_space_efficiency(deduplicated_thin_space_efficiency)
+        deduplicated_compressed_space_efficiency = config.SPACE_EFFICIENCY_DEDUPLICATED_COMPRESSED
+        self.svc.validate_supported_space_efficiency(deduplicated_compressed_space_efficiency)
+
+    def _test_test_build_kwargs_from_parameters(self, space_efficiency, pool, name, size, thin=False, compressed=False,
```
Comment (oriyarde) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:605]: test test?

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 6]
```diff
@@ -151,10 +151,20 @@ def test_create_volume_with_compressed_space_efficiency_success(self):
...
+    def test_create_volume_with_deduplicated_compressed_space_efficiency_success(self):
+        self._test_create_volume_success(config.SPACE_EFFICIENCY_DEDUPLICATED_COMPRESSED)
+        self.svc.client.svctask.mkvolume.assert_called_with(name="test_volume", unit="b", size=1024, pool="pool_name",
+                                                            deduplicated=True, compressed=True)
+
+    def test_create_volume_with_deduplicated_backward_compatibility_space_efficiency_success(self):
         self._test_create_volume_success(config.SPACE_EFFICIENCY_DEDUPLICATED)
         self.svc.client.svctask.mkvolume.assert_called_with(name="test_volume", unit="b", size=1024, pool="pool_name",
-                                                            compressed=True, deduplicated=True)
+                                                            deduplicated=True, compressed=True)
```
Comment (oriyarde) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:167]: why this change?

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 7]
```diff
@@ -587,22 +597,34 @@ def test_validate_supported_space_efficiency_success(self):
...
+                                                deduplicated=False):
+        expected_kwargs = {'name': name, 'unit': 'b', 'size': size, 'pool': pool}
+        if thin:
+            expected_kwargs.update({'thin': True})
+        if compressed:
+            expected_kwargs.update({'compressed': True})
+        if deduplicated:
+            expected_kwargs.update({'deduplicated': True})
+        returned_kwargs = build_kwargs_from_parameters(space_efficiency, pool, name, size)
+        self.assertDictEqual(returned_kwargs, expected_kwargs)
```
Comment (oriyarde) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:615]:
```suggestion
        actual_kwargs = build_kwargs_from_parameters(space_efficiency, pool, name, size)
        self.assertDictEqual(actual_kwargs, expected_kwargs)
```

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 8]
```diff
@@ -587,22 +597,34 @@ def test_validate_supported_space_efficiency_success(self):
...
+
+    def _test_test_build_kwargs_from_parameters(self, space_efficiency, pool, name, size, thin=False, compressed=False,
+                                                deduplicated=False):
+        expected_kwargs = {'name': name, 'unit': 'b', 'size': size, 'pool': pool}
+        if thin:
+            expected_kwargs.update({'thin': True})
+        if compressed:
+            expected_kwargs.update({'compressed': True})
+        if deduplicated:
+            expected_kwargs.update({'deduplicated': True})
```
Comment (oriyarde) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:613]: can't we just have a `space_efficiency_kwargs` dict parameter, to avoid this code?

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 6 above)
Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:167]: I was just rewriting all the dedup,  I didn't remember what was the old test order..

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 8 above)
Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:613]: there are 5 options (thin, compressed, deduplicated, dedup_thin, dedup_compressed), its not really better IMO..

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 7 above)
Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: done

File: controller/array_action/array_mediator_svc.py (hunk 4 above)
Comment (ArbelNathan) [reply controller/array_action/array_mediator_svc.py]: done

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 5 above)
Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: done

File: controller/array_action/config.py (hunk 1 above)
Comment (ArbelNathan) [reply controller/array_action/config.py]: done

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 8 above)
Comment (oriyarde) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:613]: I just don't think there is a need for this code

Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: done

File: controller/array_action/array_mediator_svc.py (hunk 4 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py]: I don't see the change in the function name

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 6 above)
Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: done

File: controller/array_action/array_mediator_svc.py (hunk 4 above)
Comment (ArbelNathan) [reply controller/array_action/array_mediator_svc.py]: oh, now

**Review by oriyarde — APPROVED** [review]
LGTM
